File: src/main/Java/kafka/avro/SimpleAvroConsumer.java
package kafka.avro;

import com.twitter.bijection.Injection;
import com.twitter.bijection.avro.GenericAvroCodecs;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericRecord;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.Arrays;
import java.util.Properties;

public class SimpleAvroConsumer {

    public static void main(String[] args) throws IOException {
        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("group.id", "mygroup");
        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        props.put("value.deserializer", "org.apache.kafka.common.serialization.ByteArrayDeserializer");

        String schemaFile = new String(Files.readAllBytes(Paths.get("./src/main/Java/kafka/avro/record.avsc")));

        KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<>(props);
        consumer.subscribe(Arrays.asList("avro-topic"));

        boolean running = true;
        while (running) {
            ConsumerRecords<String, byte[]> records = consumer.poll(Duration.ofMillis(100));
            for (ConsumerRecord<String, byte[]> avroRecord : records) {

                Schema.Parser parser = new Schema.Parser();
                Schema schema = parser.parse(schemaFile);
                Injection<GenericRecord, byte[]> recordInjection = GenericAvroCodecs.toBinary(schema);
                GenericRecord record = recordInjection.invert(avroRecord.value()).get();

                System.out.println("Receiving record: str1= " + record.get("str1")
                        + ", str2= " + record.get("str2")
                        + ", int1=" + record.get("int1"));
            }
        }
        consumer.close();
    }
}

File: src/main/scala/alpakka/kafka/avro/SimpleAvroConsumer.scala
package alpakka.kafka.avro

import com.sksamuel.avro4s.*
import org.apache.kafka.clients.consumer.{ConsumerRecord, KafkaConsumer}
import org.apache.kafka.common.serialization.StringDeserializer

import java.time.Duration
import java.util.Properties
import scala.jdk.CollectionConverters.*

/**
  * Not pekko streams related
  *
  * Prerequisite:
  * Run [[alpakka.env.KafkaServerEmbedded]]
  * Run [[alpakka.kafka.avro.SimpleAvroProducer]]
  */
object SimpleAvroConsumer {
  def main(args: Array[String]): Unit = {
    new Application();
    ()
  }

  private class Application {
    val props = new Properties()
    props.put("bootstrap.servers", "localhost:29092")
    props.put("group.id", "mygroup")
    props.put("key.deserializer", classOf[StringDeserializer].getName)
    props.put("value.deserializer", classOf[AvroDeserializer].getName)

    val consumer = new KafkaConsumer[String, AvroRecord](props)
    consumer.subscribe(List("avro-topic").asJava)

    var running = true
    while (running) {
      val records = consumer.poll(Duration.ofMillis(100))
      for (record: ConsumerRecord[String, AvroRecord] <- records.asScala) {
        val avroRecord = record.value()
        println(s"Receiving record: str1=${avroRecord.str1}, str2=${avroRecord.str2}, int1=${avroRecord.int1}")
      }
    }
  }
}

class AvroDeserializer extends org.apache.kafka.common.serialization.Deserializer[AvroRecord] {
  override def deserialize(topic: String, data: Array[Byte]): AvroRecord = {
    val avroSchema = AvroSchema[AvroRecord]
    val avroInputStream = AvroInputStream.binary[AvroRecord].from(data).build(avroSchema)
    val result = avroInputStream.iterator.next()
    avroInputStream.close()
    result
  }
}
